examples/scene/volume_2dtf.py

```diff
@@ -0,0 +1,217 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) Vispy Development Team. All Rights Reserved.
+# Distributed under the (new) BSD License. See LICENSE.txt for more info.
+# -----------------------------------------------------------------------------
+# vispy: gallery 2
+
+"""
+Volume Rendering with 2D Transfer Function
+================
+
+Example volume rendering with an editable 2D transfer function.
+
+The bottom view shows the 2D histogram of the data (value, gradent magnitude)
+with the transfer function overlaid.
+
+Controls:
+
+* 1  - Reset to default 1D transfer function
+* 2  - Reset to default 2D transfer function
+* 3  - Show attenuated MIP
+* 4  - Show attenuated MIP colored by depth of max value
+
+# Left click and drag to paint the transfer function
+# Alt + Left click and drag to erase on the transfer function
+
+"""
+import numpy as np
+from vispy import app, scene, io
+from vispy.color import BaseTransferFunction
+
+# Read and normalize volume data
+mri_data = np.load(io.load_data_file('brain/mri.npz'))['data']
+mri_data = np.flipud(np.rollaxis(mri_data, 1)).astype(np.float32)
+mri_data = (mri_data - mri_data.min()) / mri_data.ptp()
+
+# Prepare canvas
+canvas = scene.SceneCanvas(keys='interactive', size=(768, 1024), show=True)
+canvas.measure_fps()
+grid = canvas.central_widget.add_grid()
+
+# Set up a viewbox to display the image
+data_view = grid.add_view(row=0, col=0, row_span=2)
+# The other view will show the 2D histogram of the data with the TF over it
+cmap_view = grid.add_view(row=2, col=0, camera='panzoom')
+
+
+class GradientMagnitudeTF(BaseTransferFunction):
+    """Transfer function that uses the data value and gradient magnitude to
+    sample a 2D transfer function.
+    """
+
+    glsl_tf = """\
+    vec4 applyTransferFunction(vec4 color, vec3 loc, vec3 step) {
+        step = step / u_relative_step_size;
+        vec4 prev;
+        vec4 next;
+        // calculate normal vector from gradient
+        vec3 N; // normal
+        prev = $get_data(loc + vec3(-step.x, 0.0, 0.0));
+        next = $get_data(loc + vec3(step.x, 0.0, 0.0));
+        N[0] = colorToVal(prev) - colorToVal(next);
```

Comment:
just for consistency I'd use `normal.x` and so on instead of `N[0]`, cause I thought it was an array at first.

Reply:
Thanks - this is a good idea. To be honest I just copied this code from the iso shader (`calculateColor`) but I agree this would be more readable. It also could be just factored out as finding the gradient at a point may be otherwise useful.